diff -ru a/Config/defs.mk.in
--- a/Config/defs.mk.in	2014-12-13 18:55:54.000000000 -0500
+++ b/Config/defs.mk.in	2015-07-15 22:06:56.455125410 -0400
@@ -41,7 +41,7 @@
 exec_prefix     = @exec_prefix@
 bindir          = @bindir@
 libdir          = @libdir@
-MODDIR          = $(libdir)/$(tzsh)/$(VERSION)
+MODDIR          = $(ATHAME_TESTDIR)$(libdir)/$(tzsh)/$(VERSION)
 infodir         = @infodir@
 mandir          = @mandir@
 datarootdir     = @datarootdir@
@@ -67,7 +67,7 @@
 DLLDFLAGS       = @DLLDFLAGS@
 LIBLDFLAGS      = @LIBLDFLAGS@
 EXELDFLAGS      = @EXELDFLAGS@
-LIBS            = @LIBS@
+LIBS            = @LIBS@ -lutil
 DL_EXT          = @DL_EXT@
 DLLD            = @DLLD@
 EXPOPT          = @EXPOPT@

diff -ru a/Src/Makefile.in b/Src/Makefile.in
--- a/Src/Makefile.in	2014-12-13 18:55:54.000000000 -0500
+++ b/Src/Makefile.in	2015-07-15 22:06:56.455125410 -0400
@@ -142,8 +142,15 @@
 uninstall: uninstall.bin uninstall.modules
 .PHONY: install uninstall
 
-install.bin: install.bin-here
-uninstall.bin: uninstall.bin-here
+install.vimbed:
+	if [ -d $(DESTDIR)$(libdir)/athame_zsh ]; then rm -r $(DESTDIR)$(libdir)/athame_zsh; fi
+	mkdir -p $(DESTDIR)/$(libdir)
+	cp -r $(dir_top)/$(subdir)/vimbed/plugin $(DESTDIR)/$(libdir)/athame_zsh
+uninstall.vimbed:
+	rm -r $(DESTDIR)$(libdir)/athame_zsh
+
+install.bin: install.bin-here install.vimbed
+uninstall.bin: uninstall.bin-here uninstall.vimbed
 .PHONY: install.bin uninstall.bin

 # install binary, creating install directory if necessary
diff -ru a/Src/Makemod.in.in b/Src/Makemod.in.in
--- a/Src/Makemod.in.in	2014-12-13 18:55:54.000000000 -0500
+++ b/Src/Makemod.in.in	2015-07-15 21:52:18.235086024 -0400
@@ -42,13 +42,14 @@
 
 sdir_src      = $(sdir_top)/Src
 dir_src       = $(dir_top)/Src
+athame        = -DVIMBED_LOCATION='"$(ATHAME_TESTDIR)$(libdir)/athame_zsh"' -DATHAME_VIM_BIN='"${ATHAME_VIM_BIN}"' -DATHAME_USE_JOBS_DEFAULT='${ATHAME_USE_JOBS_DEFAULT}'
 
 # ========== COMPILATION RULES ==========
 
 DNCFLAGS =
 
-COMPILE     = $(CC) -c -I. -I$(dir_top)/Src -I$(sdir_top)/Src -I$(sdir_top)/Src/Zle -I$(sdir) $(CPPFLAGS) $(DEFS) $(CFLAGS) $(D@L@CFLAGS)
-DLCOMPILE   = $(CC) -c -I. -I$(dir_top)/Src -I$(sdir_top)/Src -I$(sdir_top)/Src/Zle -I$(sdir) $(CPPFLAGS) $(DEFS) -DMODULE $(CFLAGS) $(DLCFLAGS)
+COMPILE     = $(CC) -c -I. -I$(dir_top)/Src -I$(sdir_top)/Src -I$(sdir_top)/Src/Zle -I$(sdir) $(CPPFLAGS) $(DEFS) $(CFLAGS) $(D@L@CFLAGS) $(athame)
+DLCOMPILE   = $(CC) -c -I. -I$(dir_top)/Src -I$(sdir_top)/Src -I$(sdir_top)/Src/Zle -I$(sdir) $(CPPFLAGS) $(DEFS) -DMODULE $(CFLAGS) $(DLCFLAGS) $(athame)
 LINK        = $(CC) $(LDFLAGS) $(EXELDFLAGS) $(EXTRA_LDFLAGS) -o $@
 DLLINK      = $(DLLD) $(LDFLAGS) $(LIBLDFLAGS) $(DLLDFLAGS) -o $@
 
diff -ru a/Src/Zle/zle_main.c b/Src/Zle/zle_main.c
--- a/Src/Zle/zle_main.c	2015-05-22 18:16:13.000000000 -0400
+++ b/Src/Zle/zle_main.c	2015-07-15 21:09:04.778303048 -0400
@@ -28,6 +28,7 @@
  */
 
 #include "zle.mdh"
+#include "athame.h"
 #include "zle_main.pro"
 
 #ifdef HAVE_POLL_H
@@ -1024,6 +1026,8 @@
 void
 zlecore(void)
 {
+    athame_init(SHTTY, NULL);
+
     Keymap km;
 #if !defined(HAVE_POLL) && defined(HAVE_SELECT)
     struct timeval tv;
@@ -1111,6 +1137,7 @@
 
     region_active = 0;
     popheap();
+    athame_cleanup(0);
 }
 
 /* Read a line.  It is returned metafied. */
diff -ru a/Src/Zle/zle_keymap.c b/Src/Zle/zle_keymap.c
--- a/Src/Zle/zle_keymap.c	2015-05-22 18:16:13.000000000 -0400
+++ b/Src/Zle/zle_keymap.c	2015-08-23 00:04:00.000000000 -0400
@@ -28,6 +28,7 @@
  */
 
 #include "zle.mdh"
+#include "athame.h"
 
 /*
  * Keymap structures:
@@ -1532,7 +1533,17 @@
 static int
 getkeybuf(int w)
 {
-    int c = getbyte((long)w, NULL);
+    int c;
+    if (athame_enabled())
+    {
+#ifdef MULTIBYTE_SUPPORT
+        lastchar_wide_valid = 0;
+#endif
+        c = lastchar = STOUC(athame_loop(SHTTY));
+    }
+    else {
+        c = getbyte((long)w, NULL);
+    }
 
     if(c < 0)
 	return EOF;
diff -ru a/Src/Zle/zle.mdd b/Src/Zle/zle.mdd
--- a/Src/Zle/zle.mdd	2014-12-13 18:55:54.000000000 -0500
+++ b/Src/Zle/zle.mdd	2015-07-15 21:12:11.711644764 -0400
@@ -8,7 +8,7 @@
 objects="zle_bindings.o zle_hist.o zle_keymap.o zle_main.o \
 zle_misc.o zle_move.o zle_params.o zle_refresh.o \
 zle_thingy.o zle_tricky.o zle_utils.o zle_vi.o zle_word.o \
-textobjects.o"
+textobjects.o athame.o"
 
 headers="zle.h zle_things.h"
